src/sas/sascalc/calculator/geni.py

```diff
@@ -224,41 +205,51 @@ def _calc_Iqxy_magnetic(
     #print("mag", [v.shape for v in (x, y, rho, vol, mx, my, mz)])
     _calc_Iqxy_magnetic_helper(
         Iq, qx, qy, x, y, rho, vol, mx, my, mz,
-        cos_spin, sin_spin, dd, du, ud, uu)
+        cos_spin, sin_spin, cos_phi, sin_phi, dd, du, ud, uu)
     return Iq.reshape(shape)
 
 @njit("(" + "f8[:], "*10 + "f8, "*6 + ")")
 def _calc_Iqxy_magnetic_helper(
-        Iq, qx, qy, x, y, rho, vol, mx, my, mz, cos_spin, sin_spin,
+        Iq, qx, qy, x, y, rho, vol, mx, my, mz, cos_spin, sin_spin, cos_phi, sin_phi,
         dd, du, ud, uu):
     # Process each qx, qy
     # Note: enumerating a pair is slower than direct indexing in numba
     for k in range(len(qx)):
         qxk, qyk = qx[k], qy[k]
-        # If q is near 0 then set px and py to zero.
+        # If q is near 0 then discard intensity to zero. What should be the correct result for q=0?
+
+        #if abs(qxk) > 1.e-16 or abs(qyk) > 1.e-16:
+        norm = 1/np.sqrt(qxk**2 + qyk**2) 
         # Note: norm is computed as a separate scalar so that the numba jit
         # can figure out the proper type for perp even for q = 0
-        qsq = qxk**2 + qyk**2
-        norm = 1./qsq if qsq > 1e-16 else 0.
-        perp = norm*(qyk*mx - qxk*my)
-        px = perp*(qyk*cos_spin + qxk*sin_spin)
-        py = perp*(qyk*sin_spin - qxk*cos_spin)
-        pz = mz
-
-        ephase = vol*np.exp(1j*(qxk*x + qyk*y))
+        p_hat = np.array([sin_spin * cos_phi, sin_spin * sin_phi, cos_spin ])
+        q_hat = np.array([qxk, qyk, 0]) * norm
```

Comment:
`q=0` will result in NaN since `1/|q| → ∞` and so `0 ⋅ 1/|q|` is undefined.

Is there a neutral value that you can use at `q=0`, such as the average for `q` around the unit circle?